### PR TITLE
[Merged by Bors] - atxs: drop posiotining restriction requirement

### PR DESCRIPTION
**Pull request by dshulyak**
related https://github.com/spacemeshos/go-spacemesh/issues/3774

with this change any positioning atx is considered to be valid. additionally PubLayerId was changed to PublishEpoch. the rest are changes in tests.

i will submit part that picks publish epoch based on the current clock in a separate pr.

### CODE REVIEW

File: activation/validation.go [hunk 1]
```diff
@@ -207,27 +207,17 @@ func (*Validator) NIPostChallenge(challenge *types.NIPostChallenge, atxs atxProv
...
 	if *id == types.EmptyATXID {
 		return fmt.Errorf("empty positioning atx")
-	}
-
-	if *id == goldenATXID && !publayer.GetEpoch().IsGenesis() {
-		return fmt.Errorf("golden atx used for positioning atx in epoch %d, but is only valid in epoch 1", publayer.GetEpoch())
-	}
-
-	if *id != goldenATXID {
+	} else if *id != goldenATXID {
```
Comment (fasmat) [inline activation/validation.go]: NIT: I think without the else and inverting the if makes it easier to read:

```golang
	if *id == types.EmptyATXID {
		return fmt.Errorf("empty positioning atx")
	}
	if *id == goldenATXID {
		return nil 
	}

	posAtx, err := atxs.GetAtxHeader(*id)
	if err != nil {
		return &ErrAtxNotFound{Id: *id, source: err}
	}
	if posAtx.PublishEpoch >= pubepoch {
		return fmt.Errorf("positioning atx epoch (%v) must be before %v", posAtx.PublishEpoch, pubepoch)
	}
	return nil
```

**Comment by fasmat** [comment]
Just for my understanding: The positioning ATX is still needed for the TickHeight of the ATX and smeshers are incentivised to pick a recent one to get the highest possible TickHeight?

File: common/types/activation_test.go [hunk 2]
```diff
@@ -26,12 +29,23 @@ func TestRoundEndSerialization(t *testing.T) {
...
+	var object types.ActivationTx
+	f := fuzz.NewWithSeed(1001)
+	f.Fuzz(&object)
+
+	buf := bytes.NewBuffer(nil)
+	enc := scale.NewEncoder(buf)
+	_, err := object.EncodeScale(enc)
+	require.NoError(t, err)
+
+	epoch := types.EpochID(rand.Uint32())
```
Comment (fasmat) [inline common/types/activation_test.go]: The assigned value here is overwritten in the next line

```suggestion
	var epoch types.EpochID
```

File: common/types/activation.go [hunk 3]
```diff
@@ -228,8 +223,7 @@ func (atx *ActivationTx) MarshalLogObject(encoder log.ObjectEncoder) error {
 		encoder.AddUint64("vrf_nonce", uint64(*atx.VRFNonce))
 	}
 	encoder.AddString("coinbase", atx.Coinbase.String())
-	encoder.AddUint32("pub_layer_id", atx.PubLayerID.Uint32())
-	encoder.AddUint32("epoch", uint32(atx.PublishEpoch()))
+	encoder.AddUint32("epoch", uint32(atx.PublishEpoch))
```
Comment (fasmat) [inline common/types/activation.go]: NIT: Use the method you added 🙂 

```suggestion
	encoder.AddUint32("epoch", atx.PublishEpoch.Uint32())
```

File: common/types/epoch.go [hunk 4]
```diff
@@ -11,49 +11,53 @@ import (
...
+}
+
 // EncodeScale implements scale codec interface.
-func (l EpochID) EncodeScale(e *scale.Encoder) (int, error) {
-	n, err := scale.EncodeCompact32(e, uint32(l))
+func (e EpochID) EncodeScale(enc *scale.Encoder) (int, error) {
+	n, err := scale.EncodeCompact32(enc, uint32(e))
 	if err != nil {
 		return 0, err
 	}
```
Comment (fasmat) [inline common/types/epoch.go]: I know this is unrelated to your changes, but I think this is not correct. `n` might be 0 but it also might not be (if some bytes were written to `enc`).

File: common/types/epoch.go [hunk 5]
```diff
@@ -11,49 +11,53 @@ import (
...
 }
 
 // DecodeScale implements scale codec interface.
-func (l *EpochID) DecodeScale(d *scale.Decoder) (int, error) {
-	value, n, err := scale.DecodeCompact32(d)
+func (e *EpochID) DecodeScale(dec *scale.Decoder) (int, error) {
+	value, n, err := scale.DecodeCompact32(dec)
 	if err != nil {
 		return 0, err
 	}
```
Comment (fasmat) [inline common/types/epoch.go]: Again unrelated, but this is definitely wrong. To produce the error some bytes of the underlying `io.Reader`/`scale.Decoder` must have been consumed.

File: common/types/verified_activation.go [hunk 6]
```diff
@@ -48,8 +48,8 @@ func (vatx *VerifiedActivationTx) MarshalLogObject(encoder log.ObjectEncoder) er
 		encoder.AddUint64("vrf_nonce", uint64(*vatx.VRFNonce))
 	}
 	encoder.AddString("coinbase", vatx.Coinbase.String())
-	encoder.AddUint32("pub_layer_id", vatx.PubLayerID.Uint32())
-	encoder.AddUint32("epoch", uint32(vatx.PublishEpoch()))
+	encoder.AddUint32("pub_layer_id", vatx.PublishEpoch.Uint32())
```
Comment (fasmat) [inline common/types/verified_activation.go]: I think this line can be removed

**Comment by dshulyak** [comment]
> Just for my understanding: The positioning ATX is still needed for the TickHeight of the ATX and smeshers are incentivised to pick a recent one to get the highest possible TickHeight?

yes, afaik that part doesnt change. but now i am not sure if we should pick just the highest one, or the highest one from the latest epoch.

**Review by fasmat — APPROVED** [review]

**Comment by dshulyak** [comment]
@fasmat thank you, fixed

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**